proposals/eligibility_validator.go

```diff
@@ -162,27 +163,24 @@ func (v *Validator) validateReference(ballot *types.Ballot, owned *types.Activat
 
 // validateSecondary executed for non-reference ballots in latest epoch and all ballots in past epochs.
 func (v *Validator) validateSecondary(ballot *types.Ballot, owned *types.ActivationTxHeader) (*types.EpochData, error) {
-	var refballot *types.Ballot
 	if ballot.RefBallot == types.EmptyBallotID {
-		refballot = ballot
-	} else {
-		var err error
-		refballot, err = ballots.Get(v.cdb, ballot.RefBallot)
-		if err != nil {
-			return nil, fmt.Errorf("ref ballot is missing %v: %w", ballot.RefBallot, err)
+		if ballot.EpochData == nil {
+			return nil, fmt.Errorf("%w: ref ballot %v", errMissingEpochData, ballot.ID())
 		}
+		return ballot.EpochData, nil
 	}
-	if refballot.EpochData == nil {
-		return nil, fmt.Errorf("%w: ref ballot %v", errMissingEpochData, refballot.ID())
+	refdata := v.tortoise.GetBallot(ballot.RefBallot)
+	if refdata == nil {
+		return nil, fmt.Errorf("ref ballot is missing %v", ballot.RefBallot)
```

Comment:
in general my concern is when data is outside tortoise's window. why not query this from db in case of failure?

Reply:
if data is outside window, we can't decode ballot votes anymore. there have to be some heuristic to reload the window at that time, like a lot of atx weight wasn't counted